x/reporter/reporter.go

```diff
@@ -0,0 +1,211 @@
+// Copyright 2023 Jigsaw Operations LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package reporter
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"log"
+	"math/rand"
+	"net/http"
+	"net/url"
+	"sort"
+	"time"
+)
+
+var debugLog log.Logger = *log.New(io.Discard, "", 0)
+var httpClient = &http.Client{}
+
+type ConnectivityReport struct {
+	// Connection setup
+	Connection interface{} `json:"connection"`
+	// Observations
+	Time       time.Time   `json:"time"`
+	DurationMs int64       `json:"durationMs"`
+	Error      interface{} `json:"error"`
+}
+
+type Report interface {
+	IsSuccess() bool
+	CanMarshal() bool
+}
+
+// ConnectivityReport implements the Report interface
+func (r ConnectivityReport) IsSuccess() bool {
+	if r.Error == nil {
+		return true
+	} else {
+		return false
+	}
+}
+
+// Makes sure the Report type can be marshalled into JSON
+func (r ConnectivityReport) CanMarshal() bool {
+	_, err := json.Marshal(r)
+	if err != nil {
+		log.Printf("Error encoding JSON: %s\n", err)
+		return false
+	} else {
+		return true
+	}
+}
+
+type Collector interface {
+	Collect(Report) error
+}
+
+type RemoteCollector struct {
+	collectorEndpoint *url.URL
+}
+
+type SamplingCollector struct {
+	collector       Collector
+	successFraction float64
+	failureFraction float64
+}
+
+type CollectorTarget struct {
+	collector Collector
+	priority  int
+	maxRetry  int
```

Comment:
Create a `RetryCollector(collector, numRetries)` instead.

We will need to define an error type that allows the RetryCollector to determine whether the error is due to a bad request or not (like 4xx vx 5xx in HTTP), because bad requests are not retriable.